fairseq/progress_bar.py

```diff
@@ -46,9 +48,18 @@ def build_progress_bar(args, iterator, epoch=None, prefix=None, default='tqdm',
                 raise ImportError("fb_tbmf_wrapper package not found.")
             g_tbmf_wrapper = fb_tbmf_wrapper
         bar = g_tbmf_wrapper(bar, args, args.log_interval)
-    elif args.tensorboard_logdir and distributed_utils.is_master(args):
+    elif (
+        (
+            args.tensorboard_logdir
+            and getattr(args, 'use_gpu', True)
+            and distributed_utils.is_master(args)
+        ) or (
+            args.tensorboard_logdir
+            and not getattr(args, 'use_gpu', True)
+            and xm.is_master_ordinal()
```

Comment:
@dlibenzi , this is what we do in `test_train_imagenet` also, but wouldn't this create a different graph for the master ordinal compared to the other cores?

Reply:
Yes, code like that is wrong.
Code should be:
```Python
do_work_to_get_cpu_data()
if xm.is_master_ordinal():
  externalize_cpu_data_work()
```

Reply:
Note that is the input to the tensorboard writer is already CPU data (or python scalars), then the code is fine as it is.

Reply:
I will verify everything logged is cpu data before merging, and update PR if it's not the case.

Reply:
So, this code is actually just creating a different progress bar for the master ordinal, before the training even starts. Then, during training, that special progress bar is actually logging tensors on the device every N steps, but it's doing it during the step closure.

Here is pdb trace:

```
(Pdb) w
...
-> xm.mark_step()
  /anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch_xla/core/xla_model.py(414)mark_step()
-> _run_step_closures()
  /anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch_xla/core/xla_model.py(403)_run_step_closures()
-> closure()
  /anaconda3/envs/torch-xla-nightly/lib/python3.6/site-packages/torch_xla/core/xla_model.py(395)<lambda>()
-> step_closures.append(lambda a=args: closure(*a))
  /home/taylanbil/tpu-examples/deps/fairseq/train.py(420)print_training_update()
-> progress.log(stats, step=stats['num_updates'])
  /home/taylanbil/tpu-examples/deps/fairseq/fairseq/progress_bar.py(298)log()
-> self._log_to_tensorboard(stats, tag, step)
> /home/taylanbil/tpu-examples/deps/fairseq/fairseq/progress_bar.py(322)_log_to_tensorboard()
-> for key in stats.keys() - {'num_updates'}:
(Pdb)
```

This should be fine, right?

Reply:
As long as all the cores are doing the same "XLA tensor work", it is fine.

Reply:
One core is adding the loss tensors to the tb summary writer, and the others don't do that. But that core is doing this during the step closure. Is this still not recommended?

Reply:
Not good.
They will be doing different work.

Reply:
If should be:
```Python
def _tb_update(some_tensors...):
  cpu_tensors = some_tensors...cpu()
  tensorboard.write(cpu_tensors)


xm.add_step_closure(_tb_update, args=(some_tensors...))

```

Reply:
Ooookay, so this PR is a bit less trivial then I first thought then. I'll make sure they do absolutely the same work to avoid any issues.